tidy3d/web/api/container.py

```diff
@@ -504,41 +522,62 @@ def run(self, path_dir: str = DEFAULT_DATA_DIR) -> BatchData:
         data from file one by one. If no file exists for that task, it downloads it.
         """
+        # TODO: multi-threaded version of this (end to end)
         self._check_path_dir(path_dir)
+        self.upload()
         self.start()
         self.monitor()
+        self.download(path_dir=path_dir)
         return self.load(path_dir=path_dir)
 
-    @pd.validator("jobs", always=True)
-    def _upload(cls, val, values) -> None:
+    @cached_property
+    def jobs(self) -> Dict[TaskName, Job]:
         """Create a series of tasks in the :class:`.Batch` and upload them to server.
 
         Note
         ----
         To start the simulations running, must call :meth:`Batch.start` after uploaded.
         """
-        if val is not None:
-            return val
 
         # the type of job to upload (to generalize to subclasses)
-        JobType = cls.__fields__["jobs"].type_
-        parent_tasks = values.get("parent_tasks")
+        JobType = self._job_type
+        self_dict = self.dict()
 
-        verbose = bool(values.get("verbose"))
-        solver_version = values.get("solver_version")
+        # TODO: multi-threading
         jobs = {}
-        for task_name, simulation in values.get("simulations").items():
-            upload_kwargs = {key: values.get(key) for key in JobType._upload_fields}
-            upload_kwargs["task_name"] = task_name
-            upload_kwargs["simulation"] = simulation
-            upload_kwargs["verbose"] = verbose
-            upload_kwargs["solver_version"] = solver_version
-            if parent_tasks and task_name in parent_tasks:
-                upload_kwargs["parent_tasks"] = parent_tasks[task_name]
-            job = JobType(**upload_kwargs)
+        for task_name, simulation in self.simulations.items():
+            job_kwargs = {}
+
+            for key in JobType._upload_fields:
+                if key in self_dict:
+                    job_kwargs[key] = self_dict.get(key)
+
+            job_kwargs["task_name"] = task_name
+            job_kwargs["simulation"] = simulation
+            job_kwargs["verbose"] = False
+            job_kwargs["solver_version"] = self.solver_version
+            if self.parent_tasks and task_name in self.parent_tasks:
+                job_kwargs["parent_tasks"] = self.parent_tasks[task_name]
+            job = JobType(**job_kwargs)
             jobs[task_name] = job
         return jobs
 
+    @property
+    def num_jobs(self) -> int:
+        """Number of jobs in the batch."""
+        return len(self.jobs)
+
+    def upload(self) -> None:
+        """Upload a series of tasks associated with this ``Batch`` using multi-threading."""
+
+        if self.verbose:
+            console = get_logging_console()
+            console.log(f"Uploading data for {self.num_jobs} tasks.")
+
+        with ThreadPoolExecutor(max_workers=self.num_workers) as executor:
+            for task_name, job in self.jobs.items():
+                future = executor.submit(job.upload)
```

Comment:
We could use [`concurrent.futures.as_completed`](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.wait) to update a progress bar here only based on the number of uploads completed. At least is better feedback than nothing. The same could be used for downloads. There is an [example](https://docs.python.org/3/library/concurrent.futures.html#threadpoolexecutor-example) o fhow to use it to wait for completed futures.

Reply:
Thanks, this worked like a charm. Now I have a single progress bar for download and upload that updates based on the number of tasks uploaded / downloaded.